lib/OpenQA/Settings.pm

```diff
@@ -0,0 +1,39 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+package OpenQA::Settings;
+
+use strict;
+use warnings;
+
+# replace %NAME% with $settings{NAME}
+sub expand_placeholders {
+    my ($settings) = @_;
+
+    for my $value (values %$settings) {
```

Comment:
Should this really be done in random order? That seems a bit risky to me.

Reply:
It was done in random order before... and sorting it e.g. in ascending order is also a bit arbitrary (why not in descending order). I would assume the order doesn't matter the this algorithm.

Reply:
Ah, right. The while loop means that order doesn't matter. But circular references would cause an infinite loop.
```perl
my $settings = {
    one => '%two%',
    two => '%one%'
};
```

Reply:
That is true. Likely the old code had the same problem. Nevertheless we should detect this situation and reject the settings in this case - and there should be a test case for this.

Reply:
I modify this algorithm to avoid an infinite loop. I think my resolution is not good, but I have no idea how to improve it. I committed the code,  you could give me some suggestions. Thanks. @kraih @Martchus

Reply:
Thanks. I'll have a look at the code.

Reply:
If we need to reproduce a certain bug depending on in which order the hash keys are iterated, it could be useful to `tie` the hash to https://metacpan.org/pod/Tie::IxHash, so it's guaranteed that we see the bug if it happens again).

Reply:
It would be much easier to just `sort` the hash keys though to get a reproducible order.

Reply:
> It would be much easier to just `sort` the hash keys though to get a reproducible order.

Yeah, I personally prefer sorting things like this, too. A reproducible order makes finding bugs much easier.